gpkit/model.py

```diff
@@ -1,250 +1,424 @@
+        #if not all([isinstance(c, Posynomial) for c in constraints]):
+        #    raise ValueError("'solve()' can only be called on models"
+        #                     " that do not contain Signomials.")
+        return self._solve("gp", solver, verbosity, skipfailures, *args, **kwargs)
+
+    def localsolve(self, solver=None, verbosity=2, skipfailures=True, *args, **kwargs):
+        #if all([isinstance(c, Posynomial) for c in constraints]):
+        #    raise ValueError("'localsolve()' can only be called on models"
+        #                     " that contain Signomials.")
+        return self._solve("sp", solver, verbosity, skipfailures, *args, **kwargs)
+
+    def _solve(self, programType, solver, verbosity, skipfailures, *args, **kwargs):
         posynomials = self.posynomials
-        if not exps and not cs:
-            exps = functools_reduce(add, (x.exps for x in posynomials))
-            cs = np.hstack((mag(p.cs) for p in posynomials))
-        varlocs, varkeys = locate_vars(exps)
-
-        self.unsubbed = PosyTuple(exps, cs, varlocs, {})
-        self.variables = {k: Monomial(k) for k in varlocs}
-        self.varkeys = varkeys
-        self.load(self.unsubbed)
-
-        # k [j]: number of monomials (columns of F) present in each constraint
-        self.k = [len(p.cs) for p in posynomials]
-
-        # p_idxs [i]: posynomial index of each monomial
-        p_idx = 0
-        self.p_idxs = []
-        for p_len in self.k:
-            self.p_idxs += [p_idx]*p_len
-            p_idx += 1
-        self.p_idxs = np.array(self.p_idxs)
-
-    def sub(self, substitutions, val=None, frombase='last', replace=True):
-        """Substitutes into a model, modifying it in place.
-
-        Usage
-        -----
-        gp.sub({'x': 1, y: 2})
-        gp.sub(x, 3, replace=True)
-
-        Arguments
-        ---------
-        substitutions : dict or key
-            Either a dictionary whose keys are strings, Variables, or VarKeys, 
-            and whose values are numbers, or a string, Variable or Varkey.
-        val : number (optional)
-            If the substitutions entry is a single key, val holds the value
-        frombase : string (optional)
-            Which model state to update. By default updates the current state.
-        replace : boolean (optional, default is True)
-            Whether the substitution should only subtitute currently
-            unsubstituted values (False) or should also make replacements of
-            current substitutions (True).
-
-        Returns
-        -------
-        No return value: the model is modified in place.
-        """
-        # look for sweep variables
-        found_sweep = []
+        self.unsubbed_cs = np.hstack((mag(p.cs) for p in posynomials))
+        self.unsubbed_exps = functools_reduce(add, (p.exps for p in posynomials))
+        (self.unsubbed_varlocs,
+         self.unsubbed_varkeys) = locate_vars(self.unsubbed_exps)
```

Comment:
Nope, it's bad old code. They're used in parse_result, but that should be cleaner.